File: android/app/src/main/java/com/openland/spacex/transport/net/ThrustedSocket.kt
package com.openland.spacex.transport.net

import com.openland.spacex.utils.DispatchQueue

val CONNECTION_BUCKETS = arrayOf(1000, 5000, 30000)

class ThrustedSocket {

    val url: String
    val timeout: Int
    val queue: DispatchQueue

    var onOpen: (() -> Unit)? = null
    var onClose: (() -> Unit)? = null
    var onMessage: ((msg: String) -> Unit)? = null

    private var isClosed = false
    private val thruster: Thruster
    private var watchDog: WatchDogTimer? = null
    private var socket: DispatchedWebSocketClient? = null

    constructor(url: String, timeout: Int, queue: DispatchQueue) {
        this.url = url
        this.timeout = timeout
        this.queue = queue

        this.thruster = Thruster(CONNECTION_BUCKETS.map { ThrusterConfig(url, it) }.toTypedArray(), this.queue, this::onConnected)
    }

    private fun onConnected(socketClient: DispatchedWebSocketClient) {
        this.queue.requireQueue()

        this.watchDog = WatchDogTimer(this.timeout, this::onConnectionDied, this.queue)
        this.watchDog!!.reset()
        this.socket = socketClient

        socketClient.onClose = {
            this.onConnectionDied()
        }
        socketClient.onMessage = {
            if (!this.isClosed) {
                this.onMessage?.invoke(it)
            }
            this.watchDog?.kick()
        }

        this.onOpen?.invoke()
    }

    private fun onConnectionDied() {
        this.queue.requireQueue()
        if (this.isClosed) {
            return
        }

        this.isClosed = true

        // Stop Socket
        this.socket?.onMessage = null
        this.socket?.onClose = null
        this.socket?.onOpen = null
        this.socket?.close()
        this.socket = null

        // Stop Watch Dog
        this.watchDog?.kill()
        this.watchDog = null

        this.onClose?.invoke()
    }

    fun post(msg: String) {
        this.queue.requireQueue()

        if (this.socket != null) {
            this.socket!!.post(msg)
        } else {
            if (!this.isClosed) {
                throw Error("Socket is not connected yet")
            }
        }
    }

    fun close() {
        this.queue.requireQueue()

        if (this.isClosed) {
            return
        }
        this.isClosed = true

        // Stop Thruster
        this.thruster.close()

        // Stop Socket
        this.socket?.onMessage = null
        this.socket?.onClose = null
        this.socket?.onOpen = null
        this.socket?.close()
        this.socket = null

        // Stop Watch Dog
        this.watchDog?.kill()
        this.watchDog = null
    }
}

File: packages/openland-graphql/spacex/transport/net/ThrustedSocket.ts
import { WatchDogTimer } from './WatchDogTimer';
import { Thruster } from './Thruster';
import WebSocket from 'isomorphic-ws';

const empty = () => { /* */ };

const CONNECTION_BUCKETS = [1000, 5000, 30000];

export class ThrustedSocket {

    readonly url: string;
    readonly timeout: number;

    onopen: (() => void) | null = null;
    onclose: (() => void) | null = null;
    onmessage: ((msg: string) => void) | null = null;

    private thruster: Thruster;
    private socket: WebSocket | null = null;
    private watchDog: WatchDogTimer | null = null;
    private closed = false;

    constructor(url: string, timeout: number) {
        this.url = url;
        this.timeout = timeout;
        this.thruster = new Thruster(CONNECTION_BUCKETS.map((v) => ({ url, timeout: v })), this.onConnected);
    }

    private onConnected = (socket: WebSocket) => {
        this.watchDog = new WatchDogTimer(this.timeout, this.onConnectionDied);
        this.socket = socket;
        this.watchDog.reset();

        socket.onclose = () => {
            this.onConnectionDied();
        };
        socket.onmessage = (src) => {
            if (!this.closed) {
                if (this.onmessage) {
                    this.onmessage(src.data as string);
                }
            }
            if (this.watchDog) {
                this.watchDog.kick();
            }
        };
        if (this.onopen) {
            this.onopen();
        }
    }

    private onConnectionDied = () => {
        if (!this.closed) {
            this.closed = true;
            if (this.socket) {
                this.socket.onmessage = empty;
                this.socket.onclose = empty;
                this.socket.onopen = empty;
                try {
                    this.socket.close();
                } catch (e) {
                    // Ignore
                }
                this.socket = null;
            }
            if (this.watchDog) {
                this.watchDog.kill();
                this.watchDog = null;
            }
            if (this.onclose) {
                this.onclose();
            }
        }
    }

    send = (msg: string) => {
        if (this.socket) {
            this.socket.send(msg);
        } else {
            if (!this.closed) {
                throw Error('Socket is not connected yet');
            }
        }
    }

    close = () => {
        if (this.closed) {
            return;
        }
        this.closed = true;

        // Stop Thruster
        this.thruster.close();

        // Stop Socket
        if (this.socket) {
            this.socket.onmessage = empty;
            this.socket.onclose = empty;
            this.socket.onopen = empty;
            try {
                this.socket.close();
            } catch (e) {
                // Ignore
            }
            this.socket = null;
        }

        // Stop Watch Dog
        if (this.watchDog) {
            this.watchDog.kill();
            this.watchDog = null;
        }
    }
}